### PR TITLE
refactor: bump toolchain to 2024-07-19

### PR DESCRIPTION
**Pull request by MrCroxx**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

The rust toolchain version 1.81.0 is stable, which was branched from 2024-07-19.

Let's bump our toolchain version. Otherwise we cannot use lib whose MSRV is stable 1.81.0.

## Checklist

- [ ] I have written necessary rustdoc comments
- [ ] I have added necessary unit tests and integration tests
- [ ] I have added test labels as necessary. See [details](https://github.com/risingwavelabs/risingwave/blob/main/docs/developer-guide.md#ci-labels-guide).
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [ ] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [ ] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

If this PR includes changes that directly affect users or other significant modifications relevant to the community, kindly draft a release note to provide a concise summary of these changes. Please prioritize highlighting the impact these changes will have on users.

</details>

### CODE REVIEW

**Comment by MrCroxx** [comment]
@xxhZs 

https://github.com/risingwavelabs/risingwave/blob/29d2e1ef33fad45e8348af9d0de251bbd7ddd194/src/connector/src/sink/google_pubsub.rs#L175-L176

`GooglePubSubSinkDeliveryFuture` has more than 1 instance with opaque type. IIUC it is not acceptable.

**Comment by MrCroxx** [comment]
image

The opaque is `!Unpin`, which cannot be treat as `pub type GooglePubSubSinkDeliveryFuture = impl TryFuture<Ok = (), Error = SinkError> + Unpin + 'static;`

cc @xxhZs

**Review by BugenZhao — APPROVED** [review]

**Comment by xxchan** [comment]
As mentioned in https://github.com/risingwavelabs/risingwave/pull/17179, we might need to change all TAITs to bump toolchain... (unless https://github.com/rust-lang/rust/pull/128440 is landed)

**Comment by MrCroxx** [comment]
image


😳 FIrst time see this.

**Review by wenym1 — COMMENTED** [review]
Rest LGTM. 

Do we consider bumping to the latest toolchain by the way?

**Comment by wenym1** [comment]
> As mentioned in #17179, we might need to change all TAITs to bump toolchain... (unless [rust-lang/rust#128440](https://github.com/rust-lang/rust/pull/128440) is landed)

Actually it's not that many. I have fixed all the involved TAIT.

**Comment by MrCroxx** [comment]
https://github.com/risingwavelabs/risingwave/blob/d0f47d17ec52af2609647f3ae4bc403f85143fbf/src/frontend/src/optimizer/rule/index_selection_rule.rs#L965-L974

@chenzl25 Is this code still using? It seems a dead code.

File: src/storage/src/hummock/sstable/bloom.rs [hunk 1]
```diff
@@ -102,7 +102,7 @@ impl BloomFilterReader {
             true
         } else {
             let nbits = self.data.bit_len();
-            let delta = (h >> 17) | (h << 15);
+            let delta = h.rotate_left(15);
```
[the remark this answers is not in the document]
Comment (MrCroxx) [reply src/storage/src/hummock/sstable/bloom.rs:105]: IIUC, with u32 and 17 + 15 = 32, this operation equals to directly `rotate_left(15)`. (clippy also recommends it.

**Comment by MrCroxx** [comment]
https://buildkite.com/risingwavelabs/pull-request/builds/57645#0191dac7-247a-4bc7-9185-5d0c3993b6bb

@xxchan How to bump rust toolchain for https://github.com/risingwavelabs/clippy?

**Comment by MrCroxx** [comment]
UPDATES: dylint fixed via https://github.com/risingwavelabs/clippy/pull/1

**Review by xxchan — APPROVED** [review]
Thank you!!

Remember to bump CI image before merging, so that we don't install rust toolchain repeatedly in CI

**Comment by xxchan** [comment]
Wait, I think we can bump to a newer version. No need to stick to latest stable 🤔

**Comment by MrCroxx** [comment]
> Wait, I think we can bump to a newer version. No need to stick to latest stable 🤔

Agreed. Let me try. BTW, are you familiar with the the issue in the madsim integration test?

```
info: syncing channel updates for 'nightly-2024-07-19-x86_64-unknown-linux-gnu'
info: latest update on 2024-07-19, rust version 1.81.0-nightly (5affbb171 2024-07-18)
info: downloading component 'cargo'
error: component download failed for cargo-x86_64-unknown-linux-gnu

Caused by:
    could not rename downloaded file from '/root/.rustup/downloads/724035a7625f76dc52925fc2e8858e9d2c7ce6c4b31a8dd15bb57b9154590abb.partial' to '/root/.rustup/downloads/724035a7625f76dc52925fc2e8858e9d2c7ce6c4b31a8dd15bb57b9154590abb'

Stack backtrace:
   0: rustup::dist::manifestation::Manifestation::update
   1: rustup::dist::dist::try_update_from_dist_
   2: rustup::install::InstallMethod::install
   3: rustup::toolchain::distributable::DistributableToolchain::install
   4: rustup::config::Cfg::find_or_install_override_toolchain_or_default
   5: rustup::cli::proxy_mode::main
   6: rustup_init::main
   7: std::sys_common::backtrace::__rust_begin_short_backtrace
   8: main
   9: <unknown>
  10: __libc_start_main
  11: <unknown>
```

The instance keep failing fetching the components of the toolchain over and over again. Is it related to the CI image?

**Comment by MrCroxx** [comment]
> Let me try nightly-2024-09-10.

There are some false-positive cases with elided named lifetime.

**Comment by MrCroxx** [comment]
Let's just upgrade the toolchain to 2024-07-19. I need it for some stable crates with 1.81.0.

**Pull request merged**